include/swift/IDE/CompletionInstance.h

```diff
@@ -0,0 +1,69 @@
+//===--- CompletionInstance.h ---------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_IDE_COMPLETIONINSTANCE_H
+#define SWIFT_IDE_COMPLETIONINSTANCE_H
+
+#include "swift/Frontend/Frontend.h"
+#include "llvm/ADT/IntrusiveRefCntPtr.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/VirtualFileSystem.h"
+
+namespace swift {
+
+class CompilerInstance;
+class CompilerInvocation;
+class DiagnosticConsumer;
+
+namespace ide {
+
+/// Copy a memory buffer inserting '0' at the position of \c origBuf.
+std::unique_ptr<llvm::MemoryBuffer>
+makeCodeCompletionMemoryBuffer(const llvm::MemoryBuffer *origBuf,
+                               unsigned &Offset,
+                               llvm::StringRef bufferIdentifier);
+
+class CompletionInstance {
+  std::unique_ptr<CompilerInstance> CachedCI = nullptr;
+  bool EnableASTCaching = false;
+  unsigned MaxASTReuseCount = 100;
+  unsigned CurrentASTReuseCount = 0;
+
+  swift::CompilerInstance *
+  getReusingCompilerInstance(const swift::CompilerInvocation &Invocation,
```

Comment:
How about `getCachedCompilerInstance`  or `getReusableCompilerInstance`?  This could use a doc comment.